compiler/src/dotty/tools/dotc/core/SymDenotations.scala

```diff
@@ -1823,15 +1820,34 @@ object SymDenotations {
       packageObjCache
     }
 
-    /** Look first for members in package; if none are found look in package object */
-    override def computeNPMembersNamed(name: Name, inherited: Boolean)(implicit ctx: Context): PreDenotation = {
-      val denots = super.computeNPMembersNamed(name, inherited)
-      if (denots.exists) denots
-      else packageObj.moduleClass.denot match {
-        case pcls: ClassDenotation => pcls.computeNPMembersNamed(name, inherited)
-        case _ => denots
+    /** Looks in both the package object and the package for members. The precise algorithm
+     *  is as follows:
+     *
+     *  If this is the scala package or the package object exists but is currently completing,
+     *  look in the package first, and if nothing is found there, look in the package object second.
+     *  Otherwise, look in the package object first, and if nothing is found there, in
```

Comment:
The spec says "The package object should not define a member with the same name as one of the top-level objects or classes defined in package p. If there is a name conflict, the behavior of the program is currently undefined.", we probably want to update that at some points with the rules here.